File: python/src/coppeliasim/coppeliasim.py
from dataclasses import dataclass
from typing import Any
from .zmqRemoteApi import RemoteAPIClient
import random


@dataclass
class Motors:
    left: int
    right: int


@dataclass
class Robot:
    handle: int
    motors: Motors


@dataclass
class Targets:
    handle: int
    targets: list[list[float]]
    current_pos_index: int


TARGET_POS = [
    [1.810547113418579, -1.9869998693466187, 0.2530002295970917],
    [-0.15445302426815033, -1.960999608039856, 0.2530002295970917],
    [-0.14445288479328156, -0.9789999127388, 0.2530002295970917],
    [-0.1354527771472931, 0.96200031042099, 0.2530002295970917],
    [-0.144452765583992, 1.948000192642212, 0.2530002295970917],
    [1.7815473079681396, 1.9449999332427979, 0.2530002295970917],
    [
        1.7695471048355103,
        -0.034000199288129807,
        0.2530002295970917,
    ],
    [
        0.8573786616325378,
        -0.014516029506921768,
        0.2530002295970917,
    ],
    [
        -0.15445293486118317,
        -0.00700022280216217,
        0.2530002295970917,
    ],
    [-2.0844528675079346, -0.9789997935295105, 0.2530002295970917],
    [-2.0794517993927, -1.9540002346038818, 0.2530002295970917],
    [-2.093451738357544, 0.006999874487519264, 0.2530002295970917],
    [-2.0964515209198, 1.9259999990463257, 0.2530002295970917],
    [-1.089451551437378, 1.9279992580413818, 0.2530002295970917],
    [-2.0504512786865234, 0.9569994807243347, 0.2530002295970917],
    [0.8035488128662109, 1.9239994287490845, 0.2530002295970917],
    [1.8005484342575073, 0.9539993405342102, 0.2530002295970917],
    [1.7905484437942505, -0.9860007762908936, 0.2530002295970917],
]

INI_POS = [0.0, 0.0, +6.4145e-01]
INIT_ORI = [0.002015680753128251, -0.03632379239250599, 0.0]


class Coppeliasim:

    __client: RemoteAPIClient
    __sim: Any
    __robot: Robot
    __targets: Targets

    def __init__(self) -> None:
        self.__client = RemoteAPIClient()
        self.__sim = self.__client.getObject('sim')

        self.__robot = self.__connect_to_robot()

        self.__targets = self.__connect_to_targets()

    def __connect_to_robot(self) -> Robot:

        robot_handle = self.__sim.getObject('./cantoneira_dobrada')

        motors = Motors(
            left=self.__sim.getObject('./DynamicLeftJoint'),
            right=self.__sim.getObject('./DynamicRightJoint'),
        )

        return Robot(robot_handle, motors)

    def __connect_to_targets(self) -> Targets:

        target = self.__sim.getObject('./target')

        return Targets(handle=target, targets=TARGET_POS, current_pos_index=8)

    def get_target_distance(self) -> list[float]:
        relative_pos: list[float] = self.__sim.getObjectPosition(
            self.__targets.handle, self.__robot.handle)
        return relative_pos

    def get_target_pos(self) -> list[float]:
        return self.__sim.getObjectPosition(
            self.__targets.handle, -1)

    def get_robot_pos(self) -> list[float]:
        return self.__sim.getObjectPosition(self.__robot.handle, -1)

    def reset_robot_pos(self) -> None:
        self.__sim.setObjectPosition(
            self.__robot.handle, -1, INI_POS)
        self.__sim.setObjectOrientation(
            self.__robot.handle, -1,INIT_ORI)

    def set_robot_random_orientation(self):

        self.__sim.setObjectOrientation(
            self.__robot.handle, -1, [INIT_ORI[0], INIT_ORI[1], random.uniform(-3.14159265359, 3.14159265359)])

    def get_orientation_z(self) -> float:
        robot_euler_angles: list[float] = self.__sim.getObjectOrientation(
            self.__robot.handle, -1)

        return robot_euler_angles[2]

    def get_time(self) -> float:

        return self.__sim.getSimulationTime()

    def set_Velocity(self, left: float, right: float) -> None:

        self.__sim.setJointTargetVelocity(self.__robot.motors.left, left)
        self.__sim.setJointTargetVelocity(self.__robot.motors.right, right)

    def start_simulation(self) -> None:
        self.__sim.startSimulation()

    def stop_simulation(self) -> None:
        self.__sim.stopSimulation()

    def enable_step(self) -> None:
        self.__client.setStepping()

    def step(self) -> None:
        self.__client.step()

    def move_target(self) -> None:

        new_index = random.randrange(0, len(self.__targets.targets))

        self.__targets.current_pos_index = new_index

        self.__sim.setObjectPosition(
            self.__targets.handle, -1, self.__targets.targets[self.__targets.current_pos_index])

        pass

File: rust/smart_walker_controller/src/coppeliasimulation/coppeliasim.rs
use crate::coppeliasimulation::coppeliasim_sensors::{
    Accelerometer, DepthKinect, Gyroscope, RgbKinect,
};
use crate::coppeliasimulation::target::Target;
use crate::coppeliasimulation::Image;
use std::rc::Rc;
use zmq_remote_api::sim::Sim;
use zmq_remote_api::{sim, RemoteAPIError, RemoteApiClient, RemoteApiClientParams};

pub struct Coppeliasim {
    client: Rc<RemoteApiClient>,
    sim: Sim,
    accelerometer: Accelerometer,
    gyroscope: Gyroscope,
    motors_handle: Motors,
    rgb_kinect: RgbKinect,
    depth_kinect: DepthKinect,
    target: Target,
    robot_handle: i64,
}

struct Motors {
    left_handle: i64,
    right_handle: i64,
}

struct Kinect {
    rgb_handle: i64,
    depth_handle: i64,
}

impl Coppeliasim {
    pub fn new() -> Coppeliasim {
        Coppeliasim::connect().unwrap()
    }

    fn connect() -> Result<Coppeliasim, RemoteAPIError> {
        let client: RemoteApiClient = zmq_remote_api::RemoteApiClient::new(RemoteApiClientParams {
            host: "localhost".to_string(),
            ..RemoteApiClientParams::default()
        })
        .expect("unable to connect to localhost");

        let client = Rc::new(client);

        let sim: Sim = Sim::new(client.clone());

        let mass_block = sim.get_object(String::from("./Accelerometer/mass"), None)?;
        let force_sensor = sim.get_object(String::from("./Accelerometer/forceSensor"), None)?;

        let right_motor = sim.get_object(String::from("./DynamicRightJoint"), None)?;
        let left_motor = sim.get_object(String::from("./DynamicLeftJoint"), None)?;

        let rgb_kinect_handle = sim.get_object(String::from("./kinect/rgb"), None)?;
        let kinect_depth = sim.get_object(String::from("./kinect/depth"), None)?;

        let target_handle = sim.get_object(String::from("./target"), None)?;

        let robot_handle = sim.get_object(String::from("./metade_lateral_r"), None)?;

        let mut target = Target::new(target_handle);
        target.move_target(&sim)?;

        let mass = sim.get_object_float_param(mass_block, sim::SHAPEFLOATPARAM_MASS)?;

        let acc = Accelerometer {
            mass,
            handle: force_sensor,
        };

        let motors = Motors {
            left_handle: left_motor,
            right_handle: right_motor,
        };

        let coppelia_sim: Coppeliasim = Coppeliasim {
            sim,
            client,
            target,
            accelerometer: acc,
            motors_handle: motors,
            robot_handle,
            rgb_kinect: RgbKinect {
                handle: rgb_kinect_handle,
            },
            depth_kinect: DepthKinect {
                handle: kinect_depth,
            },
            gyroscope: Gyroscope {
                signal: String::from("GyroSensor"),
            },
        };

        Ok(coppelia_sim)
    }

    pub fn enable_step(&self) {
        self.client.set_stepping(true).unwrap();
    }

    pub fn step(&self) {
        self.client.step(true).unwrap();
    }

    pub fn get_target_distance(&self) -> Vec<f64> {
        self.sim
            .get_object_position(self.target.handle, self.robot_handle)
            .unwrap()
    }

    pub fn get_target_pos(&self) -> Vec<f64> {
        self.sim
            .get_object_position(self.target.handle, sim::HANDLE_WORLD)
            .unwrap()
    }

    pub fn get_robot_pos(&self) -> Vec<f64> {
        self.sim
            .get_object_position(self.robot_handle, sim::HANDLE_WORLD)
            .unwrap()
    }

    pub fn update_target(&mut self) {
        self.target.move_target(&self.sim).unwrap();
    }

    pub fn get_acceleration(&self) -> Vec<f32> {
        self.accelerometer.read(&self.sim).unwrap()
    }

    pub fn get_gyroscope_data(&self) -> Vec<f32> {
        self.gyroscope.read(&self.sim).unwrap()
    }

    pub fn get_rgb_kinect(&self) -> Image {
        self.rgb_kinect.read(&self.sim).unwrap()
    }

    pub fn get_depth_kinect(&self) -> Image {
        self.depth_kinect.read(&self.sim).unwrap()
    }

    pub fn get_time(&self) -> f64 {
        self.sim.get_simulation_time().unwrap()
    }
    pub fn get_orientation_z(&self) -> f64 {
        let ori = self
            .sim
            .get_object_orientation(self.robot_handle, sim::HANDLE_WORLD)
            .unwrap();
        ori[2]
    }

    pub fn start_simulation(&self) {
        self.sim.start_simulation().unwrap();
    }

    pub fn stop_simulation(&self) {
        self.sim.stop_simulation().unwrap();
    }

    pub fn set_velocity(&self, left: f64, right: f64) {
        self.sim
            .set_joint_target_velocity(self.motors_handle.left_handle, left as f64, None, None)
            .unwrap();
        self.sim
            .set_joint_target_velocity(self.motors_handle.right_handle, right as f64, None, None)
            .unwrap();
    }
}
